cmd/zoekt-mirror-gitea/main.go

```diff
@@ -0,0 +1,315 @@
+	clientOptions := []gitea.ClientOption{}
+	if *giteaURL != "" {
+		*giteaURL = "https://gitea.com/"
+	}
+
+	destDir := filepath.Join(*dest, host)
+	if err := os.MkdirAll(destDir, 0o755); err != nil {
+		log.Fatal(err)
+	}
+
+	if *token != "" {
+		content, err := os.ReadFile(*token)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		clientOptions = append(clientOptions, gitea.SetToken(string(content)))
+	}
+	client, err := gitea.NewClient(*giteaURL, clientOptions...)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	reposFilters := reposFilters{
+		topics:        topics,
+		excludeTopics: excludeTopics,
+		noArchived:    noArchived,
+	}
+	var repos []*gitea.Repository
+	if *org != "" {
+		repos, err = getOrgRepos(client, *org, reposFilters)
+	} else if *user != "" {
+		repos, err = getUserRepos(client, *user, reposFilters)
+	} else {
+		log.Printf("no user or org specified, cloning all repos.")
+		repos, err = getUserRepos(client, "", reposFilters)
+	}
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if !*forks {
+		trimmed := repos[:0]
+		for _, r := range repos {
+			if r.Fork == nil || !*r.Fork {
+				trimmed = append(trimmed, r)
+			}
+		}
+		repos = trimmed
+	}
+
+	filter, err := gitindex.NewFilter(*namePattern, *excludePattern)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	{
+		trimmed := repos[:0]
+		for _, r := range repos {
+			if filter.Include(*r.Name) {
+				trimmed = append(trimmed, r)
+			}
+		}
+		repos = trimmed
+	}
+
+	if err := cloneRepos(destDir, repos); err != nil {
+		log.Fatalf("cloneRepos: %v", err)
+	}
+
+	if *deleteRepos {
+		if err := deleteStaleRepos(*dest, filter, repos, *org+*user); err != nil {
+			log.Fatalf("deleteStaleRepos: %v", err)
+		}
+	}
+}
+
+func deleteStaleRepos(destDir string, filter *gitindex.Filter, repos []*gitea.Repository, user string) error {
+	var baseURL string
+	if len(repos) > 0 {
+		baseURL = repos[0].HTMLURL
+	} else {
+		return nil
+	}
+	u, err := url.Parse(baseURL)
+	if err != nil {
+		return err
+	}
+	u.Path = user
+
+	names := map[string]struct{}{}
+	for _, r := range repos {
+		u, err := url.Parse(r.HTMLURL)
+		if err != nil {
+			return err
+		}
+
+		names[filepath.Join(u.Host, u.Path+".git")] = struct{}{}
+	}
+	if err := gitindex.DeleteRepos(destDir, u, names, filter); err != nil {
+		log.Fatalf("deleteRepos: %v", err)
+	}
+	return nil
+}
+
+func hasIntersection(s1, s2 []string) bool {
+	hash := make(map[string]bool)
+	for _, e := range s1 {
+		hash[e] = true
+	}
+	for _, e := range s2 {
+		if hash[e] {
+			return true
+		}
+	}
+	return false
+}
+
+func filterRepositories(repos []*gitea.Repository, include []string, exclude []string, noArchived bool) (filteredRepos []*gitea.Repository) {
+	for _, repo := range repos {
+		if noArchived && *&repo.Archived {
+			continue
+		}
+		// FIXME: get topics from repos
+		// if (len(include) == 0 || hasIntersection(include, repo.Topics)) &&
+		// 	!hasIntersection(exclude, repo.Topics) {
+		// 	filteredRepos = append(filteredRepos, repo)
+		// }
```

Comment:
looks like you still need to do this?

Reply:
Whoops. I should've added a better note. The API doesn't have these fields yet, and this is left as an artifact for me in the future when the API is updated.